File: frontend/apps/ios/Pantopus/Features/Scheduling/SharedUI/SchedulingStatusPill.swift
//
//  SchedulingStatusPill.swift
//  Pantopus
//
//  Foundation (I0b) — the canonical booking/page status pill. Matches the
//  design's text-only chip grammar (booking-detail / bookings-inbox `StatusPill`):
//  a Title-case label on a tinted fill with a hairline tone border — NO leading
//  icon, fontSize 10 / weight 700, tight 3×8 padding. Tones: green for
//  confirmed / active, amber for pending / draft, red for declined / no-show,
//  neutral grey for paused / cancelled / completed / past / expired / secret /
//  unavailable / waitlisted. Tokens only.
//

import SwiftUI

/// Every booking, page, and link status the pill can render. The raw value is
/// the backend wire string; `init(backend:)` is tolerant of unknown values.
public enum SchedulingPillStatus: String, Sendable, Hashable, CaseIterable {
    case pending
    case confirmed
    case cancelled
    case declined
    case noShow = "no_show"
    case completed
    case past
    case active
    case paused
    case draft
    case secret
    case expired
    case unavailable
    case waitlisted
    case unknown

    /// Backend wire value (incl. aliases) → pill status.
    private static let aliases: [String: SchedulingPillStatus] = [
        "pending": .pending, "pending_approval": .pending, "requested": .pending,
        "confirmed": .confirmed, "approved": .confirmed, "accepted": .confirmed, "booked": .confirmed,
        "cancelled": .cancelled, "canceled": .cancelled,
        "declined": .declined, "rejected": .declined,
        "no_show": .noShow, "noshow": .noShow,
        "completed": .completed, "done": .completed,
        "past": .past,
        "active": .active, "live": .active, "published": .active,
        "paused": .paused,
        "draft": .draft,
        "secret": .secret, "private": .secret, "hidden": .secret,
        "expired": .expired,
        "unavailable": .unavailable, "full": .unavailable, "fully_booked": .unavailable,
        "waitlisted": .waitlisted, "waitlist": .waitlisted
    ]

    /// Map a backend status string (snake_case or camelCase) to a pill status.
    public init(backend raw: String) {
        let key = raw.lowercased().replacingOccurrences(of: "-", with: "_")
        self = Self.aliases[key] ?? .unknown
    }

    var label: String {
        switch self {
        case .pending: "Pending"
        case .confirmed: "Confirmed"
        case .cancelled: "Cancelled"
        case .declined: "Declined"
        case .noShow: "No-show"
        case .completed: "Completed"
        case .past: "Past"
        case .active: "Active"
        case .paused: "Paused"
        case .draft: "Draft"
        case .secret: "Private"
        case .expired: "Expired"
        case .unavailable: "Fully booked"
        case .waitlisted: "Waitlisted"
        case .unknown: "Status"
        }
    }

    var tone: Tone {
        switch self {
        case .confirmed, .active: .success
        case .pending, .draft: .warning
        case .declined, .noShow: .error
        case .cancelled, .completed, .past, .paused, .secret,
             .expired, .unavailable, .waitlisted, .unknown: .neutral
        }
    }

    enum Tone {
        case success, warning, error, neutral

        var background: Color {
            switch self {
            case .success: Theme.Color.successBg
            case .warning: Theme.Color.warningBg
            case .error: Theme.Color.errorBg
            case .neutral: Theme.Color.appSurfaceSunken
            }
        }

        var foreground: Color {
            switch self {
            case .success: Theme.Color.success
            case .warning: Theme.Color.warning
            case .error: Theme.Color.error
            case .neutral: Theme.Color.appTextSecondary
            }
        }

        /// Hairline border tint — the design draws a 1px tone-light outline
        /// around every status chip (neutral falls back to `appBorder`).
        var border: Color {
            switch self {
            case .success: Theme.Color.successLight
            case .warning: Theme.Color.warningLight
            case .error: Theme.Color.errorLight
            case .neutral: Theme.Color.appBorder
            }
        }
    }
}

/// A pill rendering a booking/page/link status with a semantic chip + glyph.
public struct SchedulingStatusPill: View {
    private let status: SchedulingPillStatus

    public init(_ status: SchedulingPillStatus) {
        self.status = status
    }

    /// Render directly from a backend status string.
    public init(status raw: String) {
        status = SchedulingPillStatus(backend: raw)
    }

    public var body: some View {
        Text(status.label)
            .font(.system(size: 10, weight: .bold))
            .foregroundStyle(status.tone.foreground)
            .lineLimit(1)
            .padding(.horizontal, Spacing.s2)
            .padding(.vertical, 3)
            .background(status.tone.background)
            .overlay(Capsule().strokeBorder(status.tone.border, lineWidth: 1))
            .clipShape(Capsule())
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(status.label)
            .accessibilityIdentifier("scheduling.statusPill.\(status.rawValue)")
    }
}

#if DEBUG
#Preview {
    ScrollView {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            ForEach(SchedulingPillStatus.allCases, id: \.self) { status in
                SchedulingStatusPill(status)
            }
        }
        .padding()
    }
    .background(Theme.Color.appBg)
}
#endif

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/scheduling/_shared/SchedulingStatusPill.kt
@file:Suppress("PackageNaming", "MatchingDeclarationName", "MagicNumber")

package app.pantopus.android.ui.screens.scheduling._shared

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * The canonical Calendarly booking / page / link status pill — the single source
 * of truth that mirrors iOS `SchedulingStatusPill`. Text-only chip grammar (NO
 * leading icon), 10sp / weight 700, tight 8×3 padding, tinted fill + 1dp
 * tone-light hairline border, capsule shape. Tones: green = confirmed / active,
 * amber = pending / draft, red = declined / no-show, neutral grey = paused /
 * cancelled / completed / past / expired / secret / unavailable / waitlisted.
 *
 * Replaces the per-screen variants (HubStatusPill, MyBookings StatusPill,
 * BookResource StatusPill, ManageBooking StatusBadge, …) so the chip stops
 * drifting across screens. testTag mirrors iOS `scheduling.statusPill.<status>`.
 */
enum class SchedulingPillStatus(val backend: String, val label: String) {
    Pending("pending", "Pending"),
    Confirmed("confirmed", "Confirmed"),
    Cancelled("cancelled", "Cancelled"),
    Declined("declined", "Declined"),
    NoShow("no_show", "No-show"),
    Completed("completed", "Completed"),
    Past("past", "Past"),
    Active("active", "Active"),
    Live("live", "Live"),
    Paused("paused", "Paused"),
    Draft("draft", "Draft"),
    Secret("secret", "Private"),
    Expired("expired", "Expired"),
    Unavailable("unavailable", "Fully booked"),
    Waitlisted("waitlisted", "Waitlisted"),
    Unknown("unknown", "Status"),
    ;

    internal val tone: PillTone
        get() =
            when (this) {
                Confirmed, Active, Live -> PillTone.Success
                Pending, Draft -> PillTone.Warning
                Declined, NoShow -> PillTone.Error
                Cancelled, Completed, Past, Paused, Secret,
                Expired, Unavailable, Waitlisted, Unknown,
                -> PillTone.Neutral
            }

    companion object {
        /** Backend wire value (incl. aliases) → pill status; tolerant of unknowns. */
        private val ALIASES: Map<String, SchedulingPillStatus> =
            mapOf(
                "pending" to Pending, "pending_approval" to Pending, "requested" to Pending,
                "confirmed" to Confirmed, "approved" to Confirmed, "accepted" to Confirmed, "booked" to Confirmed,
                "cancelled" to Cancelled, "canceled" to Cancelled,
                "declined" to Declined, "rejected" to Declined,
                "no_show" to NoShow, "noshow" to NoShow,
                "completed" to Completed, "done" to Completed,
                "past" to Past,
                "active" to Active, "live" to Live, "published" to Active,
                "paused" to Paused,
                "draft" to Draft,
                "secret" to Secret, "private" to Secret, "hidden" to Secret,
                "expired" to Expired,
                "unavailable" to Unavailable, "full" to Unavailable, "fully_booked" to Unavailable,
                "waitlisted" to Waitlisted, "waitlist" to Waitlisted,
            )

        fun fromBackend(raw: String): SchedulingPillStatus = ALIASES[raw.lowercase().replace('-', '_')] ?: Unknown
    }
}

internal enum class PillTone {
    Success,
    Warning,
    Error,
    Neutral,
    ;

    val bg: Color
        get() =
            when (this) {
                Success -> PantopusColors.successBg
                Warning -> PantopusColors.warningBg
                Error -> PantopusColors.errorBg
                Neutral -> PantopusColors.appSurfaceSunken
            }

    val fg: Color
        get() =
            when (this) {
                Success -> PantopusColors.success
                Warning -> PantopusColors.warning
                Error -> PantopusColors.error
                Neutral -> PantopusColors.appTextSecondary
            }

    /** Hairline border tint — a 1dp tone-light outline around every chip. */
    val border: Color
        get() =
            when (this) {
                Success -> PantopusColors.successLight
                Warning -> PantopusColors.warningLight
                Error -> PantopusColors.errorLight
                Neutral -> PantopusColors.appBorder
            }
}

@Composable
fun SchedulingStatusPill(
    status: SchedulingPillStatus,
    modifier: Modifier = Modifier,
) = StatusPillChip(label = status.label, tone = status.tone, backend = status.backend, modifier = modifier)

/**
 * Convenience overload: render directly from a backend status string. An
 * unrecognized wire value keeps a humanized form of the raw string
 * ("in_review" → "In review") instead of the generic "Status", mirroring the
 * per-screen chips this pill replaced (which echoed the raw value rather than
 * dropping it).
 */
@Composable
fun SchedulingStatusPill(
    status: String,
    modifier: Modifier = Modifier,
) {
    val mapped = SchedulingPillStatus.fromBackend(status)
    if (mapped != SchedulingPillStatus.Unknown) {
        SchedulingStatusPill(mapped, modifier)
    } else {
        StatusPillChip(
            label = humanizeStatus(status) ?: SchedulingPillStatus.Unknown.label,
            tone = SchedulingPillStatus.Unknown.tone,
            backend = status.lowercase().replace('-', '_').ifBlank { "unknown" },
            modifier = modifier,
        )
    }
}

/** "in_review" / "in-review" → "In review"; blank → null (falls back to "Status"). */
private fun humanizeStatus(raw: String): String? =
    raw.replace('_', ' ').replace('-', ' ').trim().ifBlank { null }?.replaceFirstChar { it.uppercase() }

@Composable
private fun StatusPillChip(
    label: String,
    tone: PillTone,
    backend: String,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(Radii.pill)
    Text(
        text = label,
        color = tone.fg,
        fontSize = 10.sp,
        fontWeight = FontWeight.Bold,
        maxLines = 1,
        modifier =
            modifier
                .testTag("scheduling.statusPill.$backend")
                .background(tone.bg, shape)
                .border(1.dp, tone.border, shape)
                .padding(horizontal = Spacing.s2, vertical = 3.dp),
    )
}
